src/backend/columnar/columnar_tableam.c

```diff
@@ -2579,3 +2598,325 @@ downgrade_columnar_storage(PG_FUNCTION_ARGS)
+	}
+
+	/* citus extension exists and has been created */
+	return true;
+}
+
+
+/*
+ * CheckCitusVersion checks whether there is a version mismatch between the
+ * available version and the loaded version or between the installed version
+ * and the loaded version. Returns true if compatible, false otherwise.
+ *
+ * As a side effect, this function also sets citusVersionKnownCompatible_Columnar global
+ * variable to true which reduces version check cost of next calls.
+ */
+bool
+CheckCitusVersion_Columnar(int elevel)
+{
+	if (citusVersionKnownCompatible_Columnar ||
+		!CitusHasBeenLoaded_Columnar() ||
+		!EnableVersionChecks_Columnar)
+	{
+		return true;
+	}
+
+	if (CheckAvailableVersion_Columnar(elevel) && CheckInstalledVersion_Columnar(elevel))
+	{
+		citusVersionKnownCompatible_Columnar = true;
+		return true;
+	}
+	else
+	{
+		return false;
+	}
+}
+
+
+/*
+ * CheckAvailableVersion compares CITUS_EXTENSIONVERSION and the currently
+ * available version from the citus.control file. If they are not compatible,
+ * this function logs an error with the specified elevel and returns false,
+ * otherwise it returns true.
+ */
+bool
+CheckAvailableVersion_Columnar(int elevel)
+{
+	if (!EnableVersionChecks_Columnar)
+	{
+		return true;
+	}
+
+	char *availableVersion = AvailableExtensionVersion_Columnar();
+
+	if (!MajorVersionsCompatible_Columnar(availableVersion, CITUS_EXTENSIONVERSION))
+	{
+		ereport(elevel, (errmsg("loaded Citus library version differs from latest "
+								"available extension version"),
+						 errdetail("Loaded library requires %s, but the latest control "
+								   "file specifies %s.", CITUS_MAJORVERSION,
+								   availableVersion),
+						 errhint("Restart the database to load the latest Citus "
+								 "library.")));
+		return false;
+	}
+
+	return true;
+}
+
+
+/*
+ * CheckInstalledVersion compares CITUS_EXTENSIONVERSION and the
+ * extension's current version from the pg_extension catalog table. If they
+ * are not compatible, this function logs an error with the specified elevel,
+ * otherwise it returns true.
+ */
+static bool
+CheckInstalledVersion_Columnar(int elevel)
+{
+	Assert(CitusHasBeenLoaded_Columnar());
+	Assert(EnableVersionChecks_Columnar);
+
+	char *installedVersion = InstalledExtensionVersion_Columnar();
+
+	if (!MajorVersionsCompatible_Columnar(installedVersion, CITUS_EXTENSIONVERSION))
+	{
+		ereport(elevel, (errmsg("loaded Citus library version differs from installed "
+								"extension version"),
+						 errdetail("Loaded library requires %s, but the installed "
+								   "extension version is %s.", CITUS_MAJORVERSION,
+								   installedVersion),
+						 errhint("Run ALTER EXTENSION citus UPDATE and try again.")));
+		return false;
+	}
+
+	return true;
+}
+
+
+/*
+ * MajorVersionsCompatible checks whether both versions are compatible. They
+ * are if major and minor version numbers match, the schema version is
+ * ignored.  Returns true if compatible, false otherwise.
+ */
```

Comment:
Seems like a lot of code to do a simple thing. I guess copying code is better than writing new code, but I am left wondering why the original code wasn't something like:

```
for(i = 0; leftVersion[i] && rightVersion[i]; i++)
{
    if (leftVersion[i] != rightVersion[i])
    {
        return false;
    }
    if (leftVersion[i] == '-')
    {
        break;
    }
}

return (leftVersion[i] == rightVersion[i]);
```

Reply:
I'll look into this but to commit for now, I think we can look at this at a later time when columnar is fully separated.